### PR TITLE
Add FilterSpan API to analyzer contexts

### LINKED ISSUE #67257: [Lightbulb Perf] Expose an optional FilterSpan public API on SyntaxTreeAnalysisContext and SemanticModelAnalysisContext
https://github.com/dotnet/roslyn/pull/67255#issue-1618807605 has the details.

> 4. SemanticModel analyzer- execute semantic model action callback on the input file. The driver was only doing so for compiler analyzer, but after this PR we do so for all semantic model analyzers. Note that the public API for semantic model action callbacks does not take a filter span for analysis, though we have this span available in the driver. We should consider exposing a new public API for an optional FilterSpan on SemanticModelAnalysisContext, which should allow SmanticModel analyzers to optimize the diagnostic computation and reporting for just the span, instead of entire file. I'll file a tracking issue for this.

> 5. SyntaxTree based syntax-only analyzer - execute callback on input tree. Driver already did this correctly. Note that the public API for SyntaxTreeAction on the DiagnosticAnalyzer does not take any filter span, so there is no further optimization possible here - we have to request full tree syntax diagnostics. I'll file a tracking issue for exposing an optional FilterSpan on SyntaxTreeAnalysisContext to help analyzers optimize the diagnostic computation and reporting to just the span.

### PR DESCRIPTION
**Pull request by mavasani**
Closes #67257

- Add `FilterSpan` public API to `SemanticModelAnalysisContext` and `SyntaxTreeAnalysisContext` types to allow analyzers to optimize diagnostic computation for lightbulb code path.
- Add span based public API overloads to `CompilationWithAnalyzers` to compute syntax diagnostics
- Updates our most expensive syntax tree action based analyzer (formatting analyzer - IDE0055) to respect this filter span.

## Performance measurements with this change
I wrote a custom standalone tool based on Roslyn NuGet packages that directly loads the compilation and invokes `CompilationWithAnalyzers.GetAnalysisResultAsync(tree, ...)` API with and without filterSpan. I am seeing pretty significant performance improvement for the formatting analyzer after this change, see screenshot below. Left side is time to compute full file formatting diagnostics, right side is time to compute span-only formatting diagnostics. 
FilterSpan

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler ping for reviews - this is a simple change which gives us very good performance improvement for lightbulb.

File: src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs [hunk 1]
```diff
@@ -1084,6 +1084,51 @@ internal async Task TestOnlyRequiredAnalyzerExecutedDuringDiagnosticComputation(
             Assert.Empty(diagnosticMap.Other);
         }
 
+        [Theory, WorkItem(67257, "https://github.com/dotnet/roslyn/issues/67257")]
+        [CombinatorialData]
+        public async Task TestFilterSpanOnContextAsync(bool testSyntaxTreeAction)
```
Comment (cston) [inline src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs:1089]: >TestFilterSpanOnContextAsync [](http://example.com/codeflow?start=26&length=28)

Just curious: What is the difference between this test and `DiagnosticAnalyzerTests.TestFilterSpanOnContextAsync`?

**Review by cston — APPROVED** [review]

File: src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs (hunk 1 above)
Comment (mavasani) [reply src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs:1089]: This tests the IDE change in this PR to pass in the filter span to the call into `CompilationWithAnalyzers.GetAnalysisResultAsync`

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs [hunk 2]
```diff
@@ -393,7 +409,25 @@ public async Task<ImmutableArray<Diagnostic>> GetAnalyzerSyntaxDiagnosticsAsync(
             VerifyTree(tree);
             VerifyExistingAnalyzersArgument(analyzers);
 
-            return await GetAnalyzerSyntaxDiagnosticsCoreAsync(tree, analyzers, cancellationToken).ConfigureAwait(false);
+            return await GetAnalyzerSyntaxDiagnosticsCoreAsync(tree, analyzers, filterSpan: null, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Returns syntax diagnostics produced by given <paramref name="analyzers"/> from analyzing the given <paramref name="tree"/>, optionally scoped to a <paramref name="filterSpan"/>.
+        /// Depending on analyzers' behavior, returned diagnostics can have locations outside the tree,
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:417]:
```suggestion
        /// Depending on analyzers' behavior, returned diagnostics can have locations outside the tree or filter span,
```

**Comment by mavasani** [comment]
Thanks @cston. @dotnet/roslyn-compiler for an additional review.

**Review by jjonescz — APPROVED** [review]

**Comment by mavasani** [comment]
Thank you @jjonescz!

**Pull request merged**